signal syntax errors to caller rather than to System.err RCHAIN-1064

dckc opened the pull request:
## Overview
Using the default error reporting from CUPS, syntax errors are only reported to stderr where other errors (e.g. process variable in name context) are signaled up the call chain and back to clients.
https://rchain.atlassian.net/browse/RCHAIN-1064

This PR overrides default syntax error handling from CUPS to treat syntax errors like other errors.

It also includes the offending token in syntax error messages.
https://rchain.atlassian.net/browse/RCHAIN-990

In RholangCLI, we now only print a stacktrace for bugs in the interpreter; for syntax errors, we just print the message.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

dckc commented:
motivation: [real-time syntax error feedback for rholang with emacs flycheck](https://drive.google.com/file/d/1uXIRaSJlkhK6dXo7Xsn4wxkQ_Havp2f2/view?usp=sharing)

using this bit of emacs config:

```
(flycheck-define-checker rholang-cli
  "Syntax checking for Rholang"
  ;; TODO: un-hard-code jar path
  :command ("java"
            "--add-opens=java.base/java.nio=ALL-UNNAMED"
            "-Dlogback.configurationFile=logback-rholang_cli.xml"
            "-jar" "/home/connolly/projects/rchain/rholang-cli/target/scala-2.12/rholangCLI-assembly-0.1.0-SNAPSHOT.jar"
            "--quiet" source-original)
  :standard-input t
  :modes (scala-mode java-mode)
  :error-patterns ((error (message (*? print)) " at " line ":" column))
  )

(add-to-list 'flycheck-checkers 'rholang-cli)
```

ArturGajowy commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/Interpreter.scala:
```diff
@@ -160,3 +159,22 @@ object Interpreter {
+      }
+    )
+  override def syntax_error(cur_token: Symbol): Unit             = ()
+  override def report_error(message: String, info: Object): Unit = ()
```
those "= ()` overrides are turning previous behavior into noops. Most likely this is wrong. If not, it requires a convincing explanation in a comment on the overrides.

dckc replied:
Yeah; I carefully reviewed the CUP docs and did all that reasoning in my head about why it should be this way and then didn't write it down. :-/

ArturGajowy commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/RholangCLI.scala:
```diff
@@ -129,7 +129,19 @@ object RholangCLI {
+          case th                  => th.printStackTrace(Console.err)
+        },
+        processTerm
+      )
```
This change will obscure any debug info there might be once those things happen.

Also, reasoning about how the exception is thrown at the place it's being caught is often wrong and quickly obsolete.

Granted, it should be up to the caller to decide what happens when there's an error. This mandates making processFile return an Either[Throwable, Unit] and doing the error handling on the call sites.

Please drop this commit. Feel free to do the `Either` change if it helps your goals.

dckc replied:
> This change will obscure any debug info there might be once those things happen.

If there's a syntax error (or file not found, or...), that's an error in the input supplied by the user. A stack trace is just going to obscure that fact.

ArturGajowy commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/RholangCLI.scala:
```diff
@@ -179,14 +182,20 @@ object RholangCLI {
   }
 
-  def evaluatePar(runtime: Runtime[Task])(par: Par)(implicit scheduler: Scheduler): Unit = {
+  def evaluatePar(runtime: Runtime[Task], quiet: Boolean = false)(
```
please make this private

ArturGajowy replied:
the 'add quiet flag' change should be a separate commit.

dckc replied:
perhaps all the changes to rholang-cli should be a separate PR

ArturGajowy replied:
separate PR or this PR but in separate commits - I'm fine either way :)

dckc commented:
@ArturGajowy how is this?

dckc commented:
I rebased to current dev.

dckc commented:
> commits 4. and 7. should be part of commit 1.

Done.

> Please also make sure to address the `Try[Try[Unit]]` thing ...

I gave it a try, but I'm not sure how to trap errors in `Interpreter.evaluate()` nor how / whether to change `waitForSuccess`. Help?

> ... in the commit introducing that code (vs a separate one).

I haven't squashed this part yet.

ArturGajowy commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/RholangCLI.scala:
```diff
@@ -119,20 +149,20 @@ object RholangCLI {
-      .fold(_.printStackTrace(Console.err), processTerm)
+      else evaluatePar(runtime, quiet)
+
+    Try(reader(fileName)).map(source => {
```
```suggestion
Try(reader(fileName)).flatMap(source => {
```

ArturGajowy replied:
Otherwise the error handling is 'swallow the exception', because `Try(whatever).map(Failure(_))` is `Success(Failure(whatever))`, which is a "'success'"

ArturGajowy commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/RholangCLI.scala:
```diff
@@ -154,31 +184,42 @@ object RholangCLI {
     val evaluatorTask =
       for {
-        _      <- Task.now(printNormalizedTerm(par))
+        _ <- Task.now(if (!quiet) {
```
The original version was wrong b/c it did side-effects when creating a "'pure'" value (which Task.now is supposed to be used fore)

```suggestion
        _ <- Task.delay(if (!quiet) {
```

ArturGajowy commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/RholangCLI.scala:
```diff
@@ -154,31 +184,42 @@ object RholangCLI {
         result <- Interpreter[Task].evaluate(runtime, par)
       } yield result
 
     waitForSuccess(evaluatorTask.runToFuture)
```
```suggestion
    Try(waitForSuccess(evaluatorTask.runToFuture)).map {
      if (!quiet) {
        printStorageContents(runtime.space.store)
      }
    }
```

as ugly as it is using map for that. We shouldn't be using Try in the first place, but that's a discussion for a different PR. Try is still better than we had.

ArturGajowy commented:
Hope this helps. Please find another reviewer so that we can merge this.

dckc commented:
I incorporated @ArturGajowy 's suggestions (the last one didn't quite compile because it didn't remove some code) and squashed the commits so that each has a single purpose.

@odeac would you like to be the 2nd reviewer?

dckc commented:
ugh... I lost track of this PR; I thought it was in. I guess I'll clean it up...

dckc commented:
bors r+

dckc commented:
bors r+

build failed in packaging after integration tests succeeded, so I suspect a transient error.
